### PR TITLE
docker configuration

### LINKED ISSUE #29: local docker config
As a developer, I'd like to be able to run `docker-compose up --build` and have a fully working course site

#### Acceptance Criteria:

- [x] `course_id` of the course to be used can be loaded from an env variable
- [x]  I can download the course from AWS by setting the standard AWS env variables listed in an example env file
- [x] I can manually override the path to `ocw-to-hugo` with a local installation
- [x] I can make changes to the theme and the results are hot-reloaded into the running site
- [x] If I stop the docker container, I can change the `course_id` in the env and load a different course without having to rebuild the image

### PR DESCRIPTION
**Pull request by gumaerc**
#### Pre-Flight checklist

- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-course-hugo-theme/issues/29

#### What's this PR do?
This PR adds a `docker` folder with the relevant configuration files to build a docker image of this application and run it for local development.

#### How should this be manually tested?
You should have a local copy of `ocw-to-hugo` to generate content for various steps to test this PR.  Read the README and the example configuration file.  Start with an empty `.env`, placed inside the `docker` folder and then do the following:

 - Use `ocw-to-hugo` in download mode to download and generate content for a course, using the `private` folder as a staging area.  For example: `node . -i private/input -o private/output -c private/courses.json --download`.  Make sure you have AWS credentials configured for access to an `open-learning-course-data` bucket and you have a single course listed in your `private/courses.json` file like: `{"courses":["18-06-linear-algebra-spring-2010"]}`.
 - Back in `ocw-course-hugo-theme`, `cd docker`
 - `docker-compose build`
 - `docker-compose up` - This should give you an empty webpage at http://localhost:1313
 - Using the course we downloaded earlier, copy the contents of the course's content folder (i.e. `ocw-to-hugo/private/output/content/courses/18-06-linear-algebra-spring-2010/*`) to `docker/hugo/content` and copy the course's data template (i.e. `ocw-to-hugo/private/output/data/courses/18-06-linear-algebra-spring-2010.json`) to `docker/hugo/data` and rename it to `course.json`.  Reload the page at http://localhost:1313 and you should be able to browse the course pages.  Afterwards, stop the dev server in your terminal with ctrl + c and remove the data we copied into `docker/hugo/content` and `docker/hugo/data`.
 - Copy the contents of `docker/example.env` to `docker/.env` and alter `OCW_TO_HUGO_PATH` to match your path to `ocw-to-hugo`.  Set `OCW_TO_HUGO_DOWNLOAD` to `false` and set `OCW_TO_HUGO_INPUT` to `/ocw-to-hugo/private/input`.  Set `OCW_COURSE_ID` to the `course_id` of the course you downloaded in step 1 (i.e. `18-06-linear-algebra-spring-2010`).  Run `docker-compose up` and ensure that you can browse the same course site you did earlier at http://localhost:1313.
 - Make sure you have AWS credentials configured in your shell in the same way you would for `ocw-to-hugo`.  Then, set `OCW_TO_HUGO_DOWNLOAD` to `true` and remove the `OCW_TO_HUGO_INPUT` variable or set it to `/ocw-data/input`.  Set `OCW_TEST_COURSE` to a course we haven't used yet, for example, `8-04-quantum-physics-i-spring-2016`.  Run `docker-compose up` and visit http://localhost:1313.  You should see the new course there.
 - Test hot reloading of the node dev server by making a change to any of the CSS and making sure it properly reloads with your changes.  For example, in `header.scss` if you change the `background-color` of `#course-banner` it should be immediately apparent: 
```
#course-banner {
  color: $white;
  background-color: $course-blue;

  a {
    text-decoration: none;
  }
}
```
 - Test hot reloading of the Hugo server by making a change to any templates, for example,  `course_home.html` if you're looking at the course home page.  You should see the page refresh and your change rendered.

### CODE REVIEW

**Comment by noisecapella** [comment]
I'm seeing this but I'm not sure why: `ERROR: Invalid interpolation format for "ocw-course-hugo-theme" option in service "services": "${OCW_TO_HUGO_PATH:-ocw-to-hugo}:/ocw-to-hugo"`

File: README.md [hunk 1]
```diff
@@ -1,10 +1,32 @@
...
+This is a Hugo theme for rendering an OCW course.  It is implemented as a Hugo Module and can be utilized following the instructions [here](https://gohugo.io/hugo-modules/use-modules/#use-a-module-for-a-theme).  An example implementation can be found [here](https://github.com/mitodl/ocw-course-hugo-starter).  Existing course data is provided by [`ocw-to-hugo`](https://github.com/mitodl/ocw-to-hugo), or you can provide your own markdown & course metadata.
+
+## running locally with docker
+
+This project includes a Docker configuration for local development in the `docker` folder.  An example environment file can be found at `docker/example.env`.  These variables only need to be set if you are sourcing existing OCW content using `ocw-to-hugo`, otherwise you should place your content in the `docker/hugo/content` folder and your course data template at `docker/hugo/data/course.json`.  Here is a description of the variables that are used:
+
+| key | type | required? | description |
+| --- | --- | --- | --- |
```
Comment (noisecapella) [inline README.md:10]: We're missing `OCW_TEST_COURSE`

File: docker/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+---
+
+services:
+  ocw-course-hugo-theme:
```
Comment (noisecapella) [inline docker/docker-compose.yml]: Can we rename to `theme` or `ocw-theme` for brevity?

File: docker/start.sh [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
```
Comment (noisecapella) [inline docker/start.sh:2]: Can we add `set -e -o pipefail` here? If ocw-to-hugo fails for example, we would want `docker-compose up` to fail too

File: docker/start.sh [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+echo "Starting services..."
+cd /theme && npm start &
```
Comment (noisecapella) [inline docker/start.sh]: Can we split the `&&` into separate lines in this file, for better readability?

**Review by noisecapella — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -1,10 +1,32 @@
...
+| `OCW_TO_HUGO_PATH` | string | No | The path to your locally checked out `ocw-to-hugo` repo |
+| `OCW_TO_HUGO_DOWNLOAD` | boolean | No | A boolean value that determines whether or not course data is downloaded from AWS |
+| `OCW_TO_HUGO_INPUT` | string | Only if `OCW_TO_HUGO_DOWNLOAD` is false | An override for the `-i` flag in `ocw-to-hugo`, used if you already have course data downloaded locally that you want to use.  Note that this path is used inside the docker container, so if you want a path relative to your local `ocw-to-hugo`, that is mounted inside the container at `/ocw-to-hugo`. |
+| `AWS_BUCKET_NAME` | string | Only if `OCW_TO_HUGO_DOWNLOAD` is true | The AWS bucket to download course data from, i.e. `open-learning-course-data-production` |
+| `AWS_REGION` | string | Only if `OCW_TO_HUGO_DOWNLOAD` is true | The AWS region to connect to |
+| `AWS_ACCESS_KEY_ID` | string | Only if `OCW_TO_HUGO_DOWNLOAD` is true | The AWS access key ID to use |
```
Comment (noisecapella) [inline README.md:17]: `ocw-to-hugo` uses `AWS_ACCESS_KEY` so we should change this to match that, or have some intermediate step to set `AWS_ACCESS_KEY` from `AWS_ACCESS_KEY_ID`

**Review by noisecapella — CHANGES_REQUESTED** [review]

**Comment by noisecapella** [comment]
Hot reloading works great. Just a few comments

File: README.md (hunk 5 above)
Comment (gumaerc) [reply README.md:17]: I'm switching this to `AWS_ACCESS_KEY` to match `ocw-to-hugo` at the moment, but I feel like that's an oversight and it should match `AWS_ACCESS_KEY_ID` if that's the standard and that change should be made in `ocw-to-hugo`

**Comment by gumaerc** [comment]
@noisecapella ready for another look

File: README.md (hunk 5 above)
Comment (noisecapella) [reply README.md:17]: Yeah, it makes sense to have the names be consistent and also match the standard. Maybe we can make an issue to do that for both projects

**Review by noisecapella — APPROVED** [review]

**Pull request merged**